File: android/crypto/presentation/src/main/java/dev/ohoussein/crypto/presentation/model/Crypto.kt
package dev.ohoussein.crypto.presentation.model

import androidx.compose.runtime.Immutable

@Immutable
data class BaseCrypto(
    val id: String,
    val name: String,
    val symbol: String,
    val imageUrl: String,
)

@Immutable
data class Crypto(
    val base: BaseCrypto,
    val price: CryptoPrice,
    val priceChangePercentIn24h: LabelValue<Double>?,
)

@Immutable
data class CryptoPrice(
    val labelValue: LabelValue<Double>,
)

@Immutable
data class LabelValue<V>(
    val value: V,
    val label: String,
)

@Immutable
data class CryptoDetails(
    val base: BaseCrypto,
    val hashingAlgorithm: String?,
    val homePageUrl: String?,
    val blockchainSite: String?,
    val mainRepoUrl: String?,
    val sentimentUpVotesPercentage: LabelValue<Double>?,
    val sentimentDownVotesPercentage: LabelValue<Double>?,
    val description: String,
)

File: CryptoAppiOS/CryptoAppiOS/Crypto/Model/Crypto.swift
import Foundation

struct BaseCrypto: Hashable {
    let id: String
    let name: String
    let symbol: String
    let imageUrl: String
}

struct Crypto: Hashable {
    let base: BaseCrypto
    let price: CryptoPrice
    let priceChangePercentIn24h: LabelValue<Double>?
}

struct CryptoPrice: Hashable {
    let labelValue: LabelValue<Double>
}

struct LabelValue<V>: Hashable where V: Hashable {
    let value: V
    let label: String
}

struct CryptoDetails: Hashable {
    let base: BaseCrypto
    let hashingAlgorithm: String?
    let homePageUrl: String?
    let blockchainSite: String?
    let mainRepoUrl: String?
    let sentimentUpVotesPercentage: LabelValue<Double>?
    let sentimentDownVotesPercentage: LabelValue<Double>?
    let description: String
}
